pandas/core/index.py

```diff
@@ -3818,6 +3842,549 @@ def _wrap_joined_index(self, joined, other):
 Int64Index._add_logical_methods()
 
 
+class RangeIndex(Int64Index):
+
+    """
+    Immutable Index implementing an monotonic range. RangeIndex is a
+    memory-saving special case of Int64Index limited to representing
+    monotonic ranges.
+
+    Parameters
+    ----------
+    start : int (default: 0)
+    stop : int (default: 0)
+    step : int (default: 1)
+    name : object, optional
+        Name to be stored in the index
+    copy : bool, default False
+        Make a copy of input if its a RangeIndex
+
+    Notes
+    -----
+    start can also be a RangeIndex, or a range object
+
+    """
+
+    _typ = 'rangeindex'
+    _engine_type = _index.Int64Engine
+
+    def __new__(cls, start=None, stop=None, step=None, name=None, dtype=None, fastpath=False, copy=False, **kwargs):
+        if fastpath:
+            return cls._simple_new(start, stop, step, name=name)
+
+        # RangeIndex / range
+        if isinstance(start, (RangeIndex, range)):
+
+            if dtype is None or is_int64_dtype(dtype):
+                pass
+            elif is_object_dtype(dtype):
```

Comment:
For instance methods on subclasses, it's important to preserve the parent class API. But that's not important for constructors -- nobody is going to call `type(index)` rather than `Index` or `Int64Index` to make a new integer index. So I don't see why it's essential to retain compatibility there.